protocols/gossipsub/src/subscription_filter.rs

```diff
@@ -0,0 +1,443 @@
+    fn can_subscribe(&mut self, _: &TopicHash) -> bool {
+        true
+    }
+}
+
+/// Allows only whitelisted subscriptions
+#[derive(Default, Clone)]
+pub struct WhitelistSubscriptionFilter(pub HashSet<TopicHash>);
+
+impl TopicSubscriptionFilter for WhitelistSubscriptionFilter {
+    fn can_subscribe(&mut self, topic_hash: &TopicHash) -> bool {
+        self.0.contains(topic_hash)
+    }
+}
+
+/// Adds a max count to a given subscription filter
+pub struct MaxCountSubscriptionFilter<T: TopicSubscriptionFilter> {
+    pub filter: T,
+    pub max_subscribed_topics: usize,
+    pub max_subscriptions_per_request: usize,
+}
+
+impl<T: TopicSubscriptionFilter> TopicSubscriptionFilter for MaxCountSubscriptionFilter<T> {
+    fn can_subscribe(&mut self, topic_hash: &TopicHash) -> bool {
+        self.filter.can_subscribe(topic_hash)
+    }
+
+    fn filter_incoming_subscriptions<'a>(
+        &mut self,
+        subscriptions: &'a [GossipsubSubscription],
+        currently_subscribed_topics: &BTreeSet<TopicHash>,
+    ) -> Result<HashSet<&'a GossipsubSubscription>, String> {
+        if subscriptions.len() > self.max_subscriptions_per_request {
+            return Err("too many subscriptions per request".into());
+        }
+        let result = self
+            .filter
+            .filter_incoming_subscriptions(subscriptions, currently_subscribed_topics)?;
+
+        use crate::types::GossipsubSubscriptionAction::*;
+
+        let mut unsubscribed = 0;
+        let mut new_subscribed = 0;
+        for s in &result {
+            let currently_contained = currently_subscribed_topics.contains(&s.topic_hash);
+            match s.action {
+                Unsubscribe => {
+                    if currently_contained {
+                        unsubscribed += 1;
+                    }
+                }
+                Subscribe => {
+                    if !currently_contained {
+                        new_subscribed += 1;
+                    }
+                }
+            }
+        }
+
+        if new_subscribed + currently_subscribed_topics.len()
+            > self.max_subscribed_topics + unsubscribed
+        {
+            return Err("too many subscribed topics".into());
+        }
+
+        Ok(result)
+    }
+}
+
+/// Combines two subscription filters
+pub struct CombinedSubscriptionFilters<T: TopicSubscriptionFilter, S: TopicSubscriptionFilter> {
+    pub filter1: T,
+    pub filter2: S,
+}
+
+impl<T, S> TopicSubscriptionFilter for CombinedSubscriptionFilters<T, S>
+where
+    T: TopicSubscriptionFilter,
+    S: TopicSubscriptionFilter,
+{
+    fn can_subscribe(&mut self, topic_hash: &TopicHash) -> bool {
+        self.filter1.can_subscribe(topic_hash) && self.filter2.can_subscribe(topic_hash)
+    }
+
+    fn filter_incoming_subscription_set<'a>(
+        &mut self,
+        subscriptions: HashSet<&'a GossipsubSubscription>,
+        _currently_subscribed_topics: &BTreeSet<TopicHash>,
```

Comment:
This seems to be used, why the `_` ?

Reply:
oh you are right, this is a leftover after a refactoring I did, will fix.